### PR TITLE
Create and Fetch Secrets

### PR DESCRIPTION
**Pull request by bbengfort**
This PR implements API endpoints for creating secrets by POST to `/v1/secrets` and GET (soon also POST) to `/v1/secrets/:token`. The PR primarily considers the JSON data structures for interacting with the server and the basic logic of secret creation, though no backend database is currently implemented. 

This PR also deeply considers interactions with the API on the command line and provides a lot of options for interaction that should communicate what happens on the front end. 

Questions:

1. How do we supply the password on fetch? Should it be in the headers or as a POST in the JSON body or as a query param or some combination of all of the above? REST semantics say this should be a GET to the detail endpoint and the password should be in the header, but we could make this more RPC like
3. Do we want the ability to copy the secret on the clipboard from the command line?
4. Should we only allow the secret to be fetched once, or should we allow multiple fetches before the expiration up to some limit set by the create API. 

TODO:

- [ ] Password handling per the answer to the question above 
- [ ] Implement database storage for the secrets
- [ ] Implement aging out the secrets

### CODE REVIEW

File: pkg/api/v1/client.go [hunk 1]
```diff
@@ -30,25 +33,124 @@ type api struct {
...
+	if req, err = http.NewRequestWithContext(ctx, method, endpoint.String(), body); err != nil {
+		return nil, fmt.Errorf("could not create request: %s", err)
+	}
+
+	// Set the headers on the request
+	req.Header.Add("User-Agent", "Whisper/1.0")
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Accept-Language", "en-US,en")
+	req.Header.Add("Accept-Encoding", "gzip, deflate, br")
+	req.Header.Add("Content-Type", "application/json")
```
Comment (rebeccabilbro) [inline pkg/api/v1/client.go:63]: I think we have converged on the idea to add password to the header.

File: pkg/api/v1/duration.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package api
+
+import (
+	"encoding/json"
+	"errors"
+	"time"
+)
+
+// Duration implements a JSON serializable time.Duration.
+type Duration time.Duration
```
Comment (rebeccabilbro) [inline pkg/api/v1/duration.go:10]: I believe Google secret manager uses the [protocol buffer duration package](https://pkg.go.dev/github.com/golang/protobuf/ptypes/duration), so I think it's a good idea to have something custom here to help us coordinate between the requests coming in from the client and the format expected on the backend.

File: pkg/secrets.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	"net/http"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	v1 "github.com/rotationalio/whisper/pkg/api/v1"
+	"github.com/rs/zerolog/log"
+)
+
+// DefaultSecretLifetime is one week after which the secret will be deleted.
+const DefaultSecretLifetime = time.Hour * 24 * 7
```
Comment (rebeccabilbro) [inline pkg/secrets.go:17]: We can probably leverage Google Secret to do the expiration; you can provide it as an `expire_time`(UTC when the Secret is scheduled to expire) or as a `ttl` (pb duration in seconds, must be more than 60).

**Review by rebeccabilbro — APPROVED** [review]
Looks good @bbengfort &mdash; I think it would be really interesting to see if we can implement this using Google Secrets (at least for v1) without any additional db...

Made a few notes based on my experience using the secretmanager package in case they are helpful.

As for the question about passwords, I think we're on the same page about adding them to the headers.

File: pkg/api/v1/duration.go (hunk 2 above)
Comment (bbengfort) [reply pkg/api/v1/duration.go:10]: Good note! I'll add a method to translate this type into pb duration.

File: pkg/secrets.go (hunk 3 above)
Comment (bbengfort) [reply pkg/secrets.go:17]: Awesome! This will definitely be a good way to manage things and I like the idea of doing Google Secrets only.

**Pull request merged**